lib/Dialect/Calyx/CalyxOps.cpp

```diff
@@ -307,29 +353,50 @@ static LogicalResult verifyComponentOp(ComponentOp op) {
                               "`clk`, `reset`, and output port `done`";
 }
 
+template <typename T>
+SmallVector<T> concat(const SmallVectorImpl<T> &a,
+                      const SmallVectorImpl<T> &b) {
+  SmallVector<T> out;
+  out.insert(out.end(), a.begin(), a.end());
+  out.insert(out.end(), b.begin(), b.end());
+  return out;
+}
+
 void ComponentOp::build(OpBuilder &builder, OperationState &result,
                         StringAttr name, ArrayRef<ComponentPortInfo> ports) {
   using namespace mlir::function_like_impl;
 
   result.addAttribute(::mlir::SymbolTable::getSymbolAttrName(), name);
 
-  SmallVector<Type, 8> portTypes;
-  SmallVector<Attribute, 8> portNames;
-  uint64_t numInPorts = 0;
+  std::pair<SmallVector<Type, 8>, SmallVector<Type, 8>> portIOTypes;
+  std::pair<SmallVector<Attribute, 8>, SmallVector<Attribute, 8>> portIONames;
+  SmallVector<Direction, 8> portDirections;
+  // Avoid using llvm::partition or llvm::sort to preserve relative ordering
+  // between individual inputs and outputs.
   for (auto &&port : ports) {
-    if (port.direction == PortDirection::INPUT)
-      ++numInPorts;
-    portNames.push_back(port.name);
-    portTypes.push_back(port.type);
+    (port.direction == Direction::Input ? portIOTypes.first
+                                        : portIOTypes.second)
+        .push_back(port.type);
+    (port.direction == Direction::Input ? portIONames.first
```

Comment:
Nit: Save this as a boolean, e.g. `bool isInput = port.direction == Direction::Input`.